### ISSUE TITLE
categorical labels on the fly

### ISSUE THREAD

**Issue by mrottmann** [keras-team/keras#9975]
Dear Keras team,

I'm currently training neural networks for semantic segmentation in Keras with tensorflow backend. The categorical (aka one hot) representation of the pixel-wise class labels for medium sized images and say 34 classes already requires a lot of memory. When generating these with np_utils.to_categorical, these one hot labels containing mostly zeros are transferred to the GPU. Is there any clever approach to transferring less data to the GPU and generating one hot representations on the GPU like on the fly for the current mini batch during training?

Cheers,
Matthias

**Comment by nuric** [comment]
You can use an [Embedding](https://keras.io/layers/embeddings/) layer to let it happen as part of the computation graph using gather operation in the backend. You just need to initialise the weights to identity and disable training. Here is a brief example:
```python
onehot = L.Embedding(num_labels, num_labels,
                     embeddings_initializer='identity',
                     trainable=False,
                     name='onehot')
embedded = onehot(some_input) # (?, input_shape, num_labels)
```

**Comment by mrottmann** [comment]
Thanks for you quick reply. I played a bit with it, but it's not clear to me how to including operations on labels into the computation graph. For the sake of simplicity here's a minimum example with MNIST.
`
from __future__ import print_function
import numpy as np
np.random.seed(0)

import tensorflow as tf

from keras.models import Model, Sequential, load_model
from keras.layers import Input, Embedding
from keras.layers.pooling import MaxPooling2D
from keras.layers.advanced_activations import LeakyReLU
from keras.layers.convolutional import Conv2D
from keras.layers.core import Dense, Activation, Dropout, Flatten
from keras.layers.merge import Add, Concatenate
from keras import regularizers
from keras.utils import np_utils
from keras.datasets import mnist

# -------------------------------------------------------------------
# parameters

nclasses = 10
batch_size = 256
img_rows, img_cols, img_chan = 28, 28, 1
input_shape = ( img_rows, img_cols, img_chan )
weightdecay = 1e-4
kernel_size = (3, 3)
filters = 16
epochs = 1

# -------------------------------------------------------------------
# load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()
N = X_train.shape[0]
X_train = X_train.astype('float32').reshape((N,)+input_shape) / 255.0
X_test = X_test.astype('float32').reshape((X_test.shape[0],)+input_shape) / 255.0

# -------------------------------------------------------------------
# model design
inp = Input(input_shape)

y = Conv2D(filters=filters, kernel_size=kernel_size, 
	  kernel_regularizer=regularizers.l2(weightdecay), strides=2, kernel_initializer='random_uniform')(inp)
y = LeakyReLU(0.1)(y)
y = MaxPooling2D()(y)
y = Flatten()(y)
y = Dense(nclasses, kernel_regularizer=regularizers.l2(weightdecay), kernel_initializer='random_uniform')(y)
y = Activation('softmax')(y)

model = Model(inputs=inp,outputs=y)
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

N = X_train.shape[0]

for e in range(epochs):
  pi = np.random.permutation(N)
  
  for i in range(int(N/batch_size)):
    X_batch = X_train[ pi[i*batch_size:(i+1)*batch_size] ]
    Y_batch = y_train[ pi[i*batch_size:(i+1)*batch_size] ]
    
    Y_batch = np_utils.to_categorical(Y_batch, nclasses) # <--- get rid of this and do it on the GPU
    
    metr    = model.train_on_batch( X_batch, Y_batch )

    print( str(i)+"/"+str(int(N/batch_size)), end=" " )
    for k in range(len(metr)):
      print( model.metrics_names[k]+": "+str(metr[k])+",", end=" " )
    print(" ")`
